### ISSUE TITLE
[STORM-1994] Add table with per-topology & worker resource usage and components in (new) supervisor and topology pages

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5776]
No Description

---
<details><summary><i>Originally reported by <a title="Alessandro Bellina" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abellina">abellina</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-1994" target="_blank">Add table with per-topology & worker resource usage and components in (new) supervisor and topology pages</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Alessandro Bellina" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abellina">abellina</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-08-22T09:36:02+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user abellina opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1994" title="Add table with per-topology & worker resource usage and components in (new) supervisor and topology pages" class="issue-link" data-issue-key="STORM-1994"><del>STORM-1994</del></a>: Add table with per-topology and worker resource usage and…</p>

<p>    … components in (new) supervisor and topology pages</p>

<p>    This adds a collapsible table to the topology page and a new supervisor page that show per worker assigned resources, number of executors and the components running on each worker. We use this as a debugging tool, helping identify where components are running and to visualize all running components in a specific host.</p>

<p>    Thanks to @kishorvpatil for coming up with the idea in the first place, and reviewing and testing the feature. Also thanks to @knusbaum, and @d2r who reviewed the code and provided bug fixes and code refactors. </p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/abellina/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/abellina/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-1994" title="Add table with per-topology & worker resource usage and components in (new) supervisor and topology pages" class="issue-link" data-issue-key="STORM-1994"><del>STORM-1994</del></a>_add_per_worker_components_and_resource_usage</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1592</p>

<hr />
<p>commit f5ad92888b6f1470e4d7f179fbe5ae73e37a07f7<br/>
Author: Alessandro Bellina <abellina@yahoo-inc.com><br/>
Date:   2016-07-06T19:23:18Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-1994" title="Add table with per-topology & worker resource usage and components in (new) supervisor and topology pages" class="issue-link" data-issue-key="STORM-1994"><del>STORM-1994</del></a>: Add table with per-topology and worker resource usage and components in (new) supervisor and topology pages</p>

<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abellina">abellina</a>:</i>
<p>I added a screenshot of the new supervisor page with the worker/component table. Also added one for the topology page counterpart.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abellina commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    Hi, I haven't gotten reviews on this ticket yet. There are a lot of generated files, so it is not as bad as it looks!</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    Hi @abellina,<br/>
    this is a bit huge patch (even without thrift generated code) including front-end change, so seems not easy for me to have time to review.</p>

<p>    Before reviewing, two considerations from me:</p>

<p>    1. Since recent RAS patches only applied to master branch, I'm not sure we may want to apply RAS related improvement to 1.x as well. Would like to hear opinions on @jerrypeng.<br/>
    2. This changes nimbus.clj which is not ported yet, and it seems not small - more than 300 lines changed only on nimbus.clj. I recently submitted a patch which modifies nimbus.clj but it only changes 11 lines.</p>

<p>    Btw, you said @kishorvpatil and @knusbaum, and @d2r already reviewed the patch. Then why not review this again (and leave +1)? Reviewing internally is no effect for Apache side.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abellina commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @HeartSaVioR, Are we freezing development on nimbus until it is ported? I can wait and submit when that is ported if nimbus is close, else I am not sure if it is a show stopper. Maybe I misunderstand your point.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @abellina <br/>
    First of all, freezing development was what we planned though it was broken now.</p>

<p>    Let me clear on this,</p>

<ul class="alternate" type="square">
	<li>Recent patches for RAS are only applied to master (2.0.0)</li>
	<li>I don't know this is for freezing development on 1.x, or just forget backporting to 1.x.</li>
	<li>At a glimpse, this patch seems to help for RAS users.</li>
	<li>Could you explain what benefits this patch give to non RAS users?</li>
	<li>So for me this patch is more likely for 2.0.0, not 1.x.</li>
	<li>If this is also for 1.x, personally I'm OK to allow modification for nimbus.clj.</li>
	<li>But if this is just for 2.0.0, porting nimbus is same milestone so why not porting it first before adding the workload for porting?</li>
</ul>


<p>    I'm also the one who works two times for same pull request because of porting, so I might be a bit keen on this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user knusbaum commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @HeartSaVioR <br/>
    I'm +1 for the patch.</p>

<p>    Nimbus has been in progress for porting for a long time. We can't (apparently) halt all progress until it's ported. You recently submitted a patch which only modified 11 lines. If many people submit patches that only modify 11 lines, it's no different. Things seem to have been pretty consistently merged regardless of the fact that they touch un-ported clojure code.</p>

<p>    This is also a pretty desirable code change. It provides something that we really need to provide if we want the platform to be successful. It's already helped me personally solve three or four issues that would have been extremely irritating to solve otherwise.</p>

<p>    If we had a time-frame for Nimbus being ported, I might say "let's wait on it," but we don't. This is a really good feature. I say, let's get it in.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abellina commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @HeartSaVioR, the patch isn't directed for RAS specifically. </p>

<p>    The primary benefit for either non-RAS or RAS is summarizing all workers for a topology in one table, and all workers for a supervisor in one table (with the new supervisor page). In each of these tables, each worker shows the components that are running (with the # of executors per component), uptime per worker, and resources assigned.</p>

<p>    The supervisor page is useful because you can go and see what other topologies are running on machine X, and what components they are running. I think that view is unique.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user jerrypeng commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @HeartSaVioR </p>

<p>    I have pushed critical bug fixes for RAS to the 1.x branch.  There has been a few algorithmic improvements on RAS that I had only put into 2.x but I desire to get them into 1.x as well when I have time.  </p>

<p>    I have found this feature to be extremely useful regardless of whether RAS is being used.  It tells you what components are scheduled in each worker and tells you the uptime of each worker so you can monitor the stability of your topology.</p>

<p>    We have no clear timeline for when the translation of nimbus will be done, and I am not sure its good to wait for an uncertain amount of time.</p>

<p>    Thus I am +1 for this feature</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @knusbaum @jerrypeng <br/>
    OK. I was just not sure that we are seeing same target version for the improvements: 1.x vs 2.x.<br/>
    If we all think we would want to ship our effort to 1.x, we must not block this due to port especially given that we all don't know when the port will be finished.<br/>
    (We must deal with this ASAP since it makes the contribution getting harder.)</p>

<p>    As I said I didn't review this, but I'm +1 for this feature.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user knusbaum commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @HeartSaVioR <br/>
    Good to hear it. I don't know if we want to stop feature development on 1.x since that's our release branch, and 2.x seems to have stalled. The policy behind all of this is probably better discussed on the mailing lists.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @jerrypeng Please let me know the issue numbers for these things. I'd also try to get some time to backport them.<br/>
    @knusbaum Agreed. I recently initiated discussion to reduce the scope for porting (let clojure tests out of phase 1) and it might be not enough.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    Applied and ran the test with -Pall-tests passed. Deployed binary distribution and test UI. Looks good. <br/>
    Assuming code review is done for other committers (two +1s). <br/>
    +1 Nice work @abellina.</p>

<p>    I would like to see configuration table as well as other page from supervisor detail page so that configuration of each machine can be easily seen, but it can be addressed to separate issue.</p>

<p>    @abellina Could we have the patch for 1.x branch, too? Or does it be easy to apply to 1.x?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abellina commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1592" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1592</a></p>

<p>    @HeartSaVioR, thanks for taking a look and running locally. As far as the supervisor config, I added: <a href="https://issues.apache.org/jira/browse/STORM-2035" class="external-link" rel="nofollow">https://issues.apache.org/jira/browse/STORM-2035</a>. Let me know if that's what you had in mind.</p>

<p>    As far as a patch for 1.x, I'll add a new PR. The stats code is in Java in 2.x, but clojure in 1.x. </p>

<p>    Thanks @knusbaum, and @jerrypeng for jumping in as well. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Thanks <a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=abellina" class="user-hover" rel="abellina">Alessandro Bellina</a> for the great work. I merged into master and 1.x branch.</p>
